File: Data Structures & Algorithms/duplicate-integer/submission-6.kt
class Solution {
    fun hasDuplicate(nums: IntArray): Boolean {
        if (nums.size < 2) {
            return false
        }
        var p1 = 0; 
        var p2 = 1;
        while (p1 != nums.size-1 || p2 != nums.size-1) {
            if (nums[p1] == nums[p2] && p1 != p2) {
                return true;
            } else {
                if (p2 == nums.size-1) {
                    p1++
                } else {
                    p2++
                }
            }
        }
        return false
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-6.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ordered = sorted(nums)
        triplets = set()
        for i, n in enumerate(ordered):
            if n > 0:
                break
            l = i+1
            r = len(ordered) - 1
            while l < r:
                if ordered[l] + ordered[r] + n == 0 and i != l and i != r:
                    t = sorted((ordered[l], ordered[r], n))
                    triplets.add(tuple(t))
                    l += 1
                    r -= 1
                elif ordered[l] + ordered[r] + n < 0:
                    l += 1
                else:
                    r -= 1
        return list(triplets)